src/server/handlers/launchpad.js

```diff
@@ -0,0 +1,236 @@
+  const parsed = parseGitHubUrl(repositoryUrl);
+  if (parsed === null || parsed.owner === null || parsed.name === null) {
+    logger.info(`Cannot parse "${repositoryUrl}"`);
+    return res.status(400).send(RESPONSE_GITHUB_BAD_URL);
+  }
+
+  const uri = `/repos/${parsed.owner}/${parsed.name}/contents/snapcraft.yaml`;
+  const options = {
+    headers: {
+      'Authorization': `token ${req.session.token}`,
+      'Accept': 'application/vnd.github.v3.raw'
+    }
+  };
+  logger.info(`Fetching snapcraft.yaml from ${repositoryUrl}`);
+  return requestGitHub.get(uri, options, (err, response, body) => {
+    if (response.statusCode !== 200) {
+      try {
+        body = JSON.parse(body);
+      } catch (e) {
+        logger.info('Invalid JSON received', err, body);
+        return res.status(500).send(RESPONSE_GITHUB_OTHER);
+      }
+      switch (body.message) {
+        case 'Not Found':
+          // snapcraft.yaml not found
+          return res.status(404).send(RESPONSE_GITHUB_NOT_FOUND);
+        case 'Bad credentials':
+          // Authentication failed
+          return res.status(401).send(RESPONSE_GITHUB_AUTHENTICATION_FAILED);
+        default:
+          // Something else
+          logger.info('GitHub API error:', err, body);
+          return res.status(500).send(RESPONSE_GITHUB_OTHER);
+      }
+    }
+
+    let snapcraftYaml;
+    try {
+      snapcraftYaml = yaml.safeLoad(body);
+    } catch (e) {
+      return res.status(400).send(RESPONSE_SNAPCRAFT_YAML_PARSE_FAILED);
+    }
+    return callback(snapcraftYaml);
+  });
+};
+
+// XXX cjwatson 2016-12-13: We should set an appropriate set of
+// architectures.
+export const newSnap = (req, res) => {
+  getSnapcraftYaml(req, res, snapcraftYaml => {
+    const repositoryUrl = req.body.repository_url;
+    if (!('name' in snapcraftYaml)) {
+      return res.status(400).send(RESPONSE_SNAPCRAFT_YAML_NO_NAME);
+    }
+    const lp_client = getLaunchpad();
+    const username = conf.get('LP_API_USERNAME');
+    logger.info(`Creating new snap for ${repositoryUrl}`);
+    lp_client.named_post('/+snaps', 'new', {
+      parameters: {
+        owner: `/~${username}`,
+        distro_series: `/${DISTRIBUTION}/${DISTRO_SERIES}`,
+        name: `${makeSnapName(repositoryUrl)}-${DISTRO_SERIES}`,
+        git_repository_url: repositoryUrl,
+        git_path: 'refs/heads/master',
+        auto_build: true,
+        auto_build_archive: `/${DISTRIBUTION}/+archive/primary`,
+        auto_build_pocket: 'Updates',
+        store_upload: true,
+        store_series: `/+snappy-series/${STORE_SERIES}`,
+        store_name: snapcraftYaml.name
+      }
+    }).then(result => {
+      logger.info(`Authorizing ${result.self_link}`);
+      lp_client.named_post(result.self_link, 'beginAuthorization', {
+        parameters: { success_url: req.body.success_url }
+      }).then(authUrl => {
+        logger.info(`Authorized ${result.self_link}`);
+        return res.redirect(authUrl);
```

Comment:
Got to this point when trying to put this into flow chart.

As far as I understand we first call for `/+snaps` `new`, and once that returns we do the SSO authorisation (based on `self_link` returned from previous call).
And the redirect url (after successful auth) is what we pass as `success_url` from request params?
This is supposed to be redirect back to BSI? Would it be directly one of our views (like /user/repo/builds?) or does auth need to redirect to some our internal endpoint to (like OAuth /verify to store the token)?